components/brave_wallet/browser/internal/hd_key_zip32.cc

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) 2024 The Brave Authors. All rights reserved.
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+#include "brave/components/brave_wallet/browser/internal/hd_key_zip32.h"
+
+#include <utility>
+
+#include "base/check.h"
+
+static_assert(BUILDFLAG(ENABLE_ORCHARD));
+
+namespace brave_wallet {
+
+HDKeyZip32::HDKeyZip32(rust::Box<OrchardExtendedSpendingKeyResult> esk)
+    : extended_spending_key_(std::move(esk)) {}
+
+HDKeyZip32::~HDKeyZip32() = default;
+
+// static
+std::unique_ptr<HDKeyZip32> HDKeyZip32::GenerateFromSeed(
+    base::span<const uint8_t> seed) {
+  auto mk = generate_orchard_extended_spending_key_from_seed(
+      rust::Slice<const uint8_t>{seed.data(), seed.size()});
+  if (mk->is_ok()) {
+    return std::make_unique<HDKeyZip32>(std::move(mk));
+  }
+  return nullptr;
+}
+
+std::unique_ptr<HDKeyZip32> HDKeyZip32::DeriveHardenedChild(uint32_t index) {
+  CHECK(extended_spending_key_->is_ok());
```

Comment:
this should not be a CHECK https://bravesoftware.slack.com/archives/C7VLGSR55/p1713599084519089

Reply:
Why do you think this shouldn't be a CHECK? If extended_spending_key is not ok that means we are working with broken HdKeyZip32 and next unwrap call will cause a crash.
Here is chromium codestyle example from the link you have provided:
https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/checks.md#examples
Similar example in the chromium code : https://source.chromium.org/chromium/chromium/src/+/main:components/user_education/common/feature_promo_session_manager.cc;l=134?q=UpdateLastActiveTime&ss=chromium
cc @iefremov

Reply:
this is a recoverable error, it does not leave the browser itself in a bad state

Reply:
I'm surprised that upstream has a CHECK for feature promo. That hardly seems like a something that should crash the entire browser over what might just be a bug in the component and I don't think Brave should crash because we somehow end up with an invalid spending key here. This is a wallet error that should be reported to the user, not a catastrophic issue that should crash the whole browser.

Reply:
This also seems like the wrong place to check this. Why aren't we checking it in the constructor?

Reply:
actually see https://github.com/brave/brave-core/pull/22870/files#r1578590482 because passing in the equivalent of `base::expected` to this class makes no sense. There should be a wrapper around the cxx code that gives us a `base::expected` with `OrchardExtendedSpendingKey` so there is no possible way to ever get an invalid instance of `OrchardExtendedSpendingKey` in the first place. If it has already been unwrapped, we know it's ok

Reply:
> I'm surprised that upstream has a CHECK for feature promo. That hardly seems like a something that should crash the entire browser over what might just be a bug in the component and I don't think Brave should crash because we somehow end up with an invalid spending key here. This is a wallet error that should be reported to the user, not a catastrophic issue that should crash the whole browser.

We already have such behavior in other places : HDKeyEd25519 crashes in the constructor and in other places you can find by searching " CHECK" in components/brave_wallet. 
Also seems chromium suggests to prefer CHECK over DCHECK in such code 
https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/checks.md#invariant_verification-mechanisms

Reply:
Moved CHECK to constructor

Reply:
> There should not be a CHECK here at all. We shouldn't crash the browser if there's a bad seed and it's completely unnecessary because you already checked this in the static method.

I was looking on this comment

> This also seems like the wrong place to check this. Why aren't we checking it in the constructor?

Ok, i just removed it so we can continue here